R/utils.R

```diff
@@ -120,14 +190,101 @@ null_transformer <- function(text, envir) {
   out
 }
 
+# Helpers for test statistics --------------------------------------
+
+# Simplify and standardize checks by grouping statistics based on variable types
+# num = numeric, bin = binary (dichotomous), mult = multinomial
+stat_types <- tibble::tribble(
+  ~resp,   ~exp,   ~stats,
+  "num",   "",     c("mean", "median", "sum", "sd", "t"),
+  "num",   "num",  c("slope", "correlation"),
+  "num",   "bin",  c("diff in means", "diff in medians", "t"),
+  "num",   "mult", c("F"),
+  "bin",   "",     c("prop", "count", "z"),
+  "bin",   "bin",  c("diff in props", "z", "ratio of props", "odds ratio", "Chisq"),
+  "bin",   "mult", c("Chisq"),
+  "mult",  "bin",  c("Chisq"),
+  "mult",  "",     c("Chisq"),
+  "mult",  "mult", c("Chisq"),
+)
+
+stat_type_desc <- tibble::tribble(
+  ~type,  ~description,
+  "num",  "numeric",
+  "bin",  "dichotomous categorical",
+  "mult", "multinomial categorical"
+)
+
+stat_desc <- tibble::tribble(
+  ~stat,               ~description,
+  "mean",              "A mean",
+  "median",            "A median", 
+  "sum",               "A sum",
+  "sd",                "A standard deviation",
+  "prop",              "A proportion", 
+  "count",             "A count", 
+  "diff in means",     "A difference in means", 
+  "diff in medians",   "A difference in medians", 
+  "diff in props",     "A difference in proportions", 
+  "Chisq",             "A chi-square statistic", 
+  "F",                 "An F statistic", 
+  "slope",             "A slope", 
+  "correlation",       "A correlation", 
+  "t",                 "A t statistic", 
+  "z",                 "A z statistic", 
+  "ratio of props",    "A ratio of proportions",  
+  "odds ratio",        "An odds ratio" 
+)
+
+implemented_stats <-  c(
+  "mean", "median", "sum", "sd", "prop", "count",
+  "diff in means", "diff in medians", "diff in props",
+  "Chisq", "F", "slope", "correlation", "t", "z",
+  "ratio of props", "odds ratio"
+)
+
+untheorized_stats <- implemented_stats[!implemented_stats %in% c(
+  "Chisq", "F", "t", "z"
+)]
+
+# Given a statistic and theory type, assume a reasonable null
+p_null <- function(x) {
+  lvls <- levels(response_variable(x))
+  num_lvls <- length(lvls)
+  probs <- 1 / num_lvls
+  
+  setNames(rep(probs, num_lvls), paste0("p.", lvls))
+}
+
+# The "null_fn" column is a function(x) whose output gives attr(x, "params")
+theorized_nulls <- tibble::tribble(
+  ~stat,    ~null_fn,
+  "Chisq", p_null,
+  "t",     function(x) {setNames(0, "mu")},
+  "z",     p_null
+)
+
+determine_variable_type <- function(x, variable) {
+  var <- eval(rlang::parse_expr(paste0(variable, "_variable(x)")))
+  
+  res <- if (is.null(var)) {
```

Comment:
It is bad practice to use assignment in multiline `if` statements. It's also redundant here.

Reply:
I'm on board. What's a good option to replace this? I gave a go at this with `switch()` but couldn't get the logic to work out. If the recommended approach here is just to assign `res <-` within each `if {}` then I'll make that happen!

Reply:
Remove `res <- ` before `if` and `res` in the end of a function body. This will lead to returning the output of the whole `if` evaluation. For me this is already OK, but I've seen that for clarity some people might add extra `return` in every `if` clause (which in this case seems redundant).

Reply:
Ah, gotcha! Thanks.